feat: 143 574 repair logs fixing time range and select visible part

raven-pan opened the pull request:
⚠️ This PR is based on [the Node Device PR](https://github.com/bigstack-oss/cube-cos-ui/pull/542), as we need to apply the latest `cube-cos-openapi` SDK to resolve Typescript type errors and  to make the CI pass.

#### What type of PR is this?

Feat

#### What this PR does / why we need it

1.) The health details table displays `raw` health events, and health time bar displays `aggregated` events.
2.) The health details table contains `fix related events`, and the health time bar displays `FIXING` time ranges.
3.) Add `NG` and `FIXING` indicators on bottom health time bar.
4.) Add a `brush` filter on bottom health time bar, allows the users to select specified time range and zoom in on top health time bar.
5.) Optimize the x-axis display by using the [d3.js ticks](https://d3js.org/d3-scale/time#time_ticks) and defining `PIXEL_PER_TICK` to make it more responsive and readable.

https://github.com/user-attachments/assets/f8bcb1ed-0c51-4cbb-886a-0fdd842dafcd

#### Which issue(s) this PR fixes

- Fixes #143 
- Fixes #574 

#### Additional documentation

#### OpenAPI

The health history was aggregated by time(10 minutes) and status(fixing related status > ng related status > ok related status). please see the `aggregate` params in the `cube-cos-openapi`:

<img width="758" height="825" alt="Screenshot 2025-07-18 at 9 00 13 PM" src="https://github.com/user-attachments/assets/91fa996d-cd7f-47b5-93f9-3984013b90a7" />

The `cube-cos-openapi` commit: https://github.com/bigstack-oss/cube-cos-openapi/pull/49/commits/cfdcdcc09041d17e34546e4f581192f557b505cf

#### Why we need to aggregate health events

1. Since we only store health events in `InfluxDB`, it's difficult to observe the system status on the time bar chart. Therefore, we convert these health event points to health status ranges.
2. Since we have a least 3 hosts (3cc), if cc1 is NG at 15:00:00 but cc2, cc3 is OK in 15:00:01, and we don't aggregate these health events , the health time bar will display only 1 second of NG, which is not reasonable.

#### How to check raw data

1. ssh to a cube host.
2. run `influx`
3. run `use events`
4. run `SELECT * FROM health LIMIT 20`

<img width="1025" height="366" alt="image" src="https://github.com/user-attachments/assets/1032dc1c-a50f-483f-ba04-ed9e286049e2" />

#### D3.js

1. Add `d3.js` package to help developers draw SVG charts.
2. Customize [d3-brush](https://d3js.org/d3-brush) to implement health time bar brush.
3. How to integrate react and d3.js: [React + D3.js](https://2019.wattenberger.com/blog/react-and-d3)

raven-pan committed:
feat: 143 574 repair logs fixing time range and select visible part

Signed-off-by: raven-pan-bigstack <raven.pan@bigstack.co>

steven-chiu-bigstack committed:
feat(web-app): device advanced actions ui

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>

steven-chiu-bigstack committed:
chore: update api version

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>

steven-chiu-bigstack committed:
refactor(web-app): move zod validateBySchema function to utils

steven-chiu-bigstack committed:
feat(web-app): integrate node device APIs

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>

steven-chiu-bigstack committed:
fix(web-app): node device ui style issues

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>

lizzy-liang-bigstack commented on packages/cube-frontend-ui-library/src/components/CosSegmentedBar/CosSegmentedBar.tsx:
```diff
@@ -115,45 +128,65 @@ export const CosSegmentedBar = (props: CosSegmentedBarProps) => {
+          x2={contentWidth}
+          strokeWidth={2}
+          stroke={cubeTheme.colors.functional['border-divider']}
+        />
```
I noticed there’s a grey line above the bar chart, but I’m not quite sure what it’s for. I tried removing it from the bar and everything still seems to work fine. Do you think we should remove it to better align with the design?

- Current Implementation
<img width="702" height="266" alt="Screenshot 2025-07-23 at 11 07 09 AM" src="https://github.com/user-attachments/assets/43f3e2b1-5977-4bbf-b184-1050281bde75" />

raven-pan replied:
Oops! that's a big mistake, thank you for pointing it out.

Fixed by this [commit](https://github.com/bigstack-oss/cube-cos-ui/pull/575/commits/2920efaca1f09dc302dd8368d51cf20b893580ba).

lizzy-liang-bigstack commented on packages/cube-frontend-web-app/src/components/HealthSegmentedBar/HealthSegmentedBar.tsx:
```diff
@@ -1,37 +1,67 @@
     width,
     segments,
+    barMarginTop: barMarginTop,
+    paddingX: paddingX,
```
Nitpick:
We might consider using shorthand syntax here for avoiding repeating variable names as both keys and values.

raven-pan replied:
Oops! that's a big mistake, thank you for pointing it out.

Fixed by this [commit](https://github.com/bigstack-oss/cube-cos-ui/pull/575/commits/2920efaca1f09dc302dd8368d51cf20b893580ba).

lizzy-liang-bigstack reviewed (approved):
Hi @raven-pan,
  
Thank you for the update! I learned a lot from your PR, especially how you used `d3.js` to handle the TimeBar and its filter logic. I also found the use of lodash/fp's `pipe` function really insightful, such a simple and elegant way to group and sort data! 🚀
  
I’ve left a few minor comments below. Let’s also check if @steven-chiu-bigstack has any additional thoughts.

Regarding the UX improvement we discussed in the meeting the other day, adding a clearer indicator on the HealthBrushTimeBar to show users that the filter can be used to zoom in/out, feel free to let us know if there’s any update from the designers. Otherwise, I think it’s good to go!

raven-pan committed:
fix: remove CosSegmentedBar redundant grey line

Signed-off-by: raven-pan-bigstack <raven.pan@bigstack.co>

raven-pan committed:
fix: health status type error

We should separate the overall health status type from the service health status type.

* Overall health status: Displayed at the top of the Health Page, includes ok and ng.
* Service health status: Displayed on the health segmented bar, includes ok, ng, and fixing.

Signed-off-by: raven-pan-bigstack <raven.pan@bigstack.co>

raven-pan committed:
refactor(health): rename `NgIndicators` to `HealthIndicators`

Since the Indicators also include fixing indicators, so rename it to HealthIndicators.

Signed-off-by: raven-pan-bigstack <raven.pan@bigstack.co>

steven-chiu-bigstack committed:
fix(web-app): node details OSD reweight value changes in edit mode issue

Signed-off-by: steven-chiu-bigstack <steven.chiu@bigstack.co>

raven-pan committed:
fix(health): svg rect negative width cause error log in browser console

Signed-off-by: raven-pan-bigstack <raven.pan@bigstack.co>

raven-pan committed:
docs(health): explain health time bar brush error message and we can ignore it for now

Signed-off-by: raven-pan-bigstack <raven.pan@bigstack.co>

raven-pan commented:
⚠️ This PR is based on https://github.com/bigstack-oss/cube-cos-ui/pull/542, as we need to apply the latest cube-cos-openapi SDK to resolve Typescript type errors and to make the CI pass.

raven-pan commented:
> Hi @raven-pan,
> 
> Thank you for the update! I learned a lot from your PR, especially how you used `d3.js` to handle the TimeBar and its filter logic. I also found the use of lodash/fp's `pipe` function really insightful, such a simple and elegant way to group and sort data! 🚀
> 
> I’ve left a few minor comments below. Let’s also check if @steven-chiu-bigstack has any additional thoughts.
> 
> Regarding the UX improvement we discussed in the meeting the other day, adding a clearer indicator on the HealthBrushTimeBar to show users that the filter can be used to zoom in/out, feel free to let us know if there’s any update from the designers. Otherwise, I think it’s good to go!

Hi @lizzy-liang-bigstack

Regarding  the suggestion to add a clearer indicator on the HealthBrushTimeBar to show users that the filter can be used to zoom in/out, I created a UI/UX enhancement ticket to track it: https://github.com/orgs/bigstack-oss/projects/6/views/1?pane=issue&itemId=121766632&issue=bigstack-oss%7Ccube-cos-ui%7C582.

Thanks!
